### PR TITLE
add math module for sequences

### PR DESCRIPTION
**Pull request by jlp765**
Not sure if this is should be a nimble package, or as a lib module.
(I couldn't find where this had already been done.  Is this superfluous?)

### CODE REVIEW

**Comment by dom96** [comment]
I don't think there is any reason not to put this in the ``math`` module.

**Comment by petermora** [comment]
The PR is lifting functions like ``sin``, ``cos`` to sequences, so ``sin(@[1.0, 2.0, 3.0])`` is working and resulting ``@[sin(1.0), sin(2.0), sin(3.0)]``. Currently these calls can be implemented as ``mapIt(@[1.0, 2.0, 3.0], sin(it))`` or ``map(@[1.0, 2.0, 3.0], x=>sin(x))``. However, the shorter form introduced by this PR (``sin(mySeq)``) is similar to what is available in Matlab, R and numPy. I can't judge whether it should be part of the lib module, but let me share some ideas for improvement.

Each function is accepting an ``openArray``, but the result is not preallocated. This leads to unnecessary allocations and copies. I suggest the form:
```nim
...
  result = newSeq[float](len(x))
  for i in 0..<x.len: result[i] = f(x[i])
```
The actual forms of the functions are very similar. A ``template`` could be used to generate them:
```nim
template liftProc(fname, inType: expr) =
  var temp: inType
  type outType = type(fname(temp))
  proc fname(x: openarray[inType]): seq[outType] =
    result = newSeq[outType](x.len)
    for i in 0..<x.len:
      result[i] = fname(x[i])

liftProc(isPowerOfTwo, int)
liftProc(sin, float)

echo isPowerOfTwo(@[1,2,3,4,5,6])
echo sin(@[1.0,2.0,3.0,4.0,5.0,6.0])
```
Using a template would generate much smaller code. I don't know how to generate the same doc string, I would ask how to access an existing doc string and return it on the forum.

I would also lift `` `+`, `-`, and, or, ... `` functions as well, so ``5+@[10, 20, 30] == @[15, 25, 35]`` and ``@[5,10] < @[6, 7, 8] == @[true, false]``. I would start with some research, have an initial implementation, ask for feedback on the forum.

By the way, if we are here (sorry, can't stop), then Mathematica can do some really cool stuff. Let's assume that you have a ``seq[seq[int]]`` and you want to transpose it: ``var transposed = transpose(orig)``. Then ``orig[x][y] == transposed[y][x]`` holds. Mathematica can do it at any level. If you have ``seq[seq[seq[seq[seq[int]]]]]``, then with the ``transpose`` function you can replace level 2 with 4.

Just some ideas.

 Peter

**Comment by jlp765** [comment]
@dom96, do you want to wait until I implement @petermora's suggestions before adding this PR?

**Comment by dom96** [comment]
@jlp765 Yes, but instead of using a template to generate each proc just use the ``sequtils.mapIt`` template.

**Comment by jlp765** [comment]
To handle the nested sequences, don't we still require a recursive template, 
and if so, then mapIt() is slower if it is as follows:


```Nim
import sequtils, math

template liftProcX(fname) =
  proc fname[T](x: openarray[T]): auto =
    var temp: T
    type outType = type(fname(temp))
    result = newSeq[outType](x.len)
    for i in 0..<x.len:
      result[i] = fname(x[i])

template liftProcY(fname) =
  proc fname[T](x: openarray[T]): auto =
    var temp: T
    type outType = type(fname(temp))
    type elType = type(x[0])
    result = x.mapIt(outType, fname(elType(it)))

liftProcX(sin)
#liftProcY(sin)
for i in 1..500000:
  var z = sin(@[@[@[i.float,2.0],@[3.0,4.0],@[5.0,6.0]], @[@[0.1,0.2],@[0.3,0.4],@[0.5,0.6]]])
```

**Comment by Varriount** [comment]
What's the difference between these procedures and using a `map` or `mapIt` with procedures in the math module?

**Comment by jlp765** [comment]
The `liftProcY` uses `mapIt()` 
The question is how to apply `mapIt()` to generically handle any `seq[seq[seq[...]]]`, hence the lifting of `mapIt()`.

When I compare the timing of `liftProcX` versus the timing of `liftProcY`, it is quicker just lifting the math function rather than lifting `mapit()`.

But I am probably not doing it the best way :-)

**Secondly**, is the issue of how best to document this, because the lifting template doesn't produce a standard proc statement that appears in the auto-generated html document via `nim doc2 ...`

**thirdly**, I am thinking (contrary to @dom96) that it is probably best NOT to put this in the `math` module, but in a separate math library, so it doesn't interfere with anyone using a linear algebra library that is doing similar but different sequence-related maths.  If they want it, they can import it specifically, but won't get it as part of the main math module.

**Comment by Araq** [comment]
> I am thinking (contrary to @dom96) that it is probably best NOT to put this in the math module, but in a separate math library, so it doesn't interfere with anyone using a linear algebra library that is doing similar but different sequence-related maths. If they want it, they can import it specifically, but won't get it as part of the main math module.

Agreed. Working on seq[float] is not common anyway, most people importing math don't need to have it work on seqs out of the box.

**Comment by jlp765** [comment]
## Round 2
Used lifting template (not mapIt) to provide nested sequences with math procs from system and math modules

Using Pythons `numpy` library as a guide, tried to add some (most?) functionality to match. (haven't added histograms, correlation, covariance, ...)

It is aimed at "Joe Average" programmers who want to do stuff to sequences of data.  It is not aimed at being a linalg/vector/matrix/.... module

### Issues:
- needs testing on Linux/Mac/win 32bit (there are asserts at the bottom)
- it hasn't lifted operators or procs that take two parameters (`+`, `-`, `mod`, `div`, ...) because couldn't work out how to do it (another challenge @petermora)
- has a mix of financial, cumulative, nested sequence lifting, 2-dim seq stuff, ...  (should this be split apart into multiple modules?)
- do you still want it in stdlib, or is this now a nimble candidate ?
- documenting the lifted procs is non-standard (because it is via a template)
- names of procs follows numpy on the non-lifted procs (but should they be called something else that is more "Nim"-ish?)
- ad. infinitum, ad. nauseum

File: lib/pure/seqmath.nim [hunk 1]
```diff
@@ -0,0 +1,1136 @@
+#
+#
+#            Nim's Runtime Library
+#        (c) Copyright 2015 Andreas Rumpf
```
Comment (Araq) [inline lib/pure/seqmath.nim]: Btw, I don't hold copyright of your work. ;-)

File: lib/pure/seqmath.nim [hunk 2]
```diff
@@ -0,0 +1,1136 @@
...
+
+  Point* = object  ## a point in 2-dimensional space, with ``x`` and ``y`` coordinates
+                   ## used with the ``bezier()`` proc
+    x*: float
+    y*: float
+
+template toFloat(f: float): float = f
+
+# ----------- Point -----------------------
+proc `!=`*[T:Point](a, b: T): bool {.inline.} =
```
Comment (Araq) [inline lib/pure/seqmath.nim]: Do not implement != or > or >=, system.nim takes care of that.

File: lib/pure/seqmath.nim [hunk 3]
```diff
@@ -0,0 +1,1136 @@
...
+liftScalarProc(floor)
+liftScalarProc(ceil)
+liftScalarProc(degToRad)
+liftScalarProc(radToDeg)
+liftScalarProc(gcd)
+liftScalarProc(lcm)
+
+# ----------- cumulative seq math -----------------------
+
+proc cumProd*[T:int|float](x: openArray[T]): seq[T] =
```
Comment (Araq) [inline lib/pure/seqmath.nim]: The T:int/float everywhere seems dangerous. I bet many algorithms work with e.g. bignums just fine. ;-)

File: lib/pure/seqmath.nim [hunk 4]
```diff
@@ -0,0 +1,1136 @@
...
+
+proc sort[T:int|float](x: openarray[T]): seq[T] =
+  ## return a sorted copy of ``x`` according to the
+  ## element type ``T``
+  result = newSeq[T](x.len)
+  for i in 0..<x.len:
+    result[i] = x[i]
+  sort(result, cmp[T])
+
+proc percentile*[T:int|float](x: openArray[T], p: int, interp = "linear"): float =
```
Comment (Araq) [inline lib/pure/seqmath.nim]: Just use an enum instead of a string. Nim is not Python. ;-)

File: lib/pure/seqmath.nim [hunk 5]
```diff
@@ -0,0 +1,1136 @@
...
+  ##  # so a net loss of $129.60 (not a good project!!)
+  var netPV = cashflows[0]
+  pv[0] = cashflows[0]
+  for i in 1..<cashflows.len:
+    pv[i] = - pv(rate, i, 0.0, cashflows[i])
+    netPV += pv[i]
+  result = netPV
+
+# ---------------------------------------------
+if isMainModule:
```
Comment (Araq) [inline lib/pure/seqmath.nim]: Use 'when isMainModule'.

File: lib/pure/seqmath.nim [hunk 6]
```diff
@@ -0,0 +1,1136 @@
...
+  assert(Point(x:2.0, y:2.0) != Point(x:2.0001, y:2.0) )
+  assert(Point(x:2.0, y:2.0) == Point(x:2.0, y:2.0) )
+  assert(Point(x:2.0, y:2.0) > Point(x:1.99999, y:2.0) )
+  assert(Point(x:1.99998, y:2.0) < Point(x:1.99999, y:2.0) )
+  assert(Point(x:1.99999, y:2.0) <= Point(x:1.99999, y:2.0) )
+  assert(Point(x:1.99999, y:2.0) >= Point(x:1.99999, y:2.0) )
+  assert(swap(Point(x:1.99999, y:2.0)) == Point(x:2.0, y:1.99999) )
+  assert(swap(Point(x:1.99999, y:2.0)) == Point(x:2.0, y:1.99999) )
```
Comment (Araq) [inline lib/pure/seqmath.nim]: I prefer 'doAssert' for tests so they are still in effect when testing things with -d:release.

**Comment by Araq** [comment]
I really like it, but I think it should be a Nimble package.

**Comment by jlp765** [comment]
Will make this a nimble package seqmath

**Pull request closed without merging**